### ISSUE TITLE
Bug: [restrict-template-expressions] Using allowNumber also allows booleans and nullish

### ISSUE THREAD

**Issue by fabio-colella-md** [typescript-eslint/typescript-eslint#11462]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAhgJwOYwLwwIwCYDMBuAKFElkSSzRigQFcBTQ46eZHSsGgG08fGYFsISDJQAGZSgBIA3mQC%2BAWhlksi5azmjeJGIPJiJ6dSgA%2BJmAHIAXnQQgLmvEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aFfKsu-RAW2LwAhn3oAPYh2TJKAeyaoMAbXDYciaNDnRIAGjXqs67JGHx4cgO4A5WPwBGm9GC6xEh7AF9PAXTU%2BXkA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const arg = 123;
const arg2 = true;
const arg3 = null;
const msg1 = `arg = ${arg}-${arg2}-${arg3}`;
const msg2 = `arg = ${arg || 'zero'}`;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      {
        "allowNumber": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Line 4, where I use `null` and `boolean` in the expression should show an error.
```ts
const msg1 = `arg = ${arg}-${arg2}-${arg3}`;
```

### Actual Result

Line 4, where I use `null` and `boolean` show no error.
```ts
const msg1 = `arg = ${arg}-${arg2}-${arg3}`;
```

**Comment by fabio-colella-md** [comment]
Tinkering in the playground it seems that this also happens with

```json
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": [
      "error", {}
    ]
  }
}
```

or 

```json
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": []
  }
}
```

**Comment by kirkwaiblinger** [comment]
Hi @fabio-colella-md !

That's because the default options include `allowNullish: true` and `allowBoolean: true` (and `allowNumber: true`)...

https://typescript-eslint.io/rules/restrict-template-expressions/#options

**Comment by fabio-colella-md** [comment]
> Hi [@fabio-colella-md](https://github.com/fabio-colella-md) !
> 
> That's because the default options include `allowNullish: true` and `allowBoolean: true` (and `allowNumber: true`)...
> 
> https://typescript-eslint.io/rules/restrict-template-expressions/#options

Right, but this also happens when using the strict type-checked version, that defaults to them being enabled 🤔

**Comment by kirkwaiblinger** [comment]
Oh, oh, I think I see now. The complaint is that this would be expected to have the errors that the strict config would normally have

[Playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=MYewdgzgLgBAhgJwOYwLwwIwCYDMBuAKFElkSSzRigQFcBTQ46eZHSsGgG08fGYFsISDJQAGZSgBIA3mQC%2BAWhlksi5azmjeJGIPJiJ6dSgA%2BJmAHIAXnQQgLmvEA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2pADgDYCuA5gJZKYAC8AnvtGgMYBO5%2B8AtE4ZfAHo08ds271GXZgAtozANbQUkALoAacFFbFCTLGFAQIkWgyZsO3XvwGsmI8mK6IAtkQCGiHrHx20acgB7JAxsGFZWQNZIDSMoU0YWdk4eND4kQSRArmIkJDkmNHdWOilglHJ4IKR9SECAM3qYzWME82SrNJssnKRiNCUuADdi0KgGps0AXxApoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Code

```ts
const arg = 123;
const arg2 = true;
const arg3 = null;
const msg1 = `arg = ${arg}-${arg2}-${arg3}`;
const msg2 = `arg = ${arg || 'zero'}`;
```

### Eslint config

```json
{
  "extends": ["plugin:@typescript-eslint/strict-type-checked"],
  "rules": {
    "@typescript-eslint/restrict-template-expressions": "error",
    "@typescript-eslint/no-unnecessary-condition": "off",
    "@typescript-eslint/no-unused-vars": "off"
  }
}
```


(Ditto for `"@typescript-eslint/restrict-template-expressions": "error"` or `"@typescript-eslint/restrict-template-expressions": ["error", {}]` or `"@typescript-eslint/restrict-template-expressions": ["error", { allowNumber: false }]`)

**Comment by fabio-colella-md** [comment]
> Oh, oh, I think I see now. The complaint is that this would be expected to have the errors that the strict config would normally have

Yes, exactly! Thanks 😊

**Comment by kirkwaiblinger** [comment]
Sadly, I think this is a limitation of the eslint config system 🙁 At runtime, the rule only sees what the resolved config is, which is a "last entry wins" system (even, apparently, if only the severity is specified). So, configs specifying custom options seemingly won't have those respected/merged if any subsequent entry exists in the config, and I don't think we have any way of working around that (short of creating a separate `restrict-template-expressions` rule with separate default options) 🫤 

@typescript-eslint/triage-team

**Comment by fabio-colella-md** [comment]
Oh, I see, so it's not merging the initial options with my custom ones but replacing them entirely 🤔 
I didn't expect that, but thanks anyway for taking a look into this and explaining the cause 😊 🙏

**Comment by kirkwaiblinger** [comment]
Wonder if we could produce expected behavior with the eslint [`settings`](https://eslint.org/docs/latest/use/configure/configuration-files#configuring-shared-settings) actually 🤔 Something like `settings: { "@typescript-eslint/use-strict-default-options": true }`

**Comment by fabio-colella-md** [comment]
That sounds promising, but I'm not entirely sure of the inner of ESLint to say if that's going to work 🤔

**Comment by bradzacher** [comment]
This is just how ESLint configs work yeah -- it works for the recommended set because the recommended set just uses the defaults. But any other config overrides the options and so if you configure the rule locally you lose the config's setup.

The same thing would happen if you, for example, used something like `eslint-config-airbnb` and then tried to change the severity of one of the rules that it enables.

My 2c: this is an expected "quirk" of ESLint and one the user must work with if they choose to locally overwrite the shared config. I don't think we should be building a system to work around this quirk in ESLint itself.
